### DISCUSSION TITLE
Error bag is empty

### DISCUSSION THREAD

**Discussion by lava83** [livewire/livewire#4230] [7. Bugs (archived)] [answered]
Hi 👋,

I have an issue here that I can't understand. In a very simple component:

### Component Class

```php
use Livewire\Component;

class ImportStockItemsComponent extends Component
{
    public $importFile;

    protected $rules = [
        'importFile' => 'required|max:1024',
    ];

    public function render(): string
    {
        return view('LiveWire.Views.import');
    }

    public function import(): void
    {
        logger('Import clicked');
        $this->validate();
    }
}
```

### View

```blade
<div>
    Errors: 
    {{ $errors }}
    <input type="file" wire:model="importFile">
    @error('importFile') {{ $message }} @enderror
    <button wire:click="import">Import</button>
</div>
```

![Bildschirmfoto 2021-11-17 um 14 42 51](https://user-images.githubusercontent.com/16590655/142211755-8258834e-35db-495e-96aa-869321fc8d20.png)

After an incorrect validation the error bag is empty. Or it can not be read during rendering. If I look at the JSON object in the DEV bar:

![Bildschirmfoto 2021-11-17 um 14 44 18](https://user-images.githubusercontent.com/16590655/142211977-57f3d069-19be-4f82-ba63-a1b75a23da18.png)

the error object is filled. Quite strange. I did some debugging and found out that `View::render` is executed twice?

**Comment by zepfietje** [comment]
Hey @lava83, have you tried wrapping the `{{ $errors }}` and `@error` lines in another element like `<div>`? It might be a DOM diffing issue: https://laravel-livewire.com/docs/2.x/troubleshooting.

**Comment by joshhanley** [comment, marked as answer]
@lava83 I've just tested your code locally and it runs happily for me.

What patch version of Livewire are you using? I tested this on v2.7.2.

![recording (51)](https://user-images.githubusercontent.com/882837/142277180-1dee7044-313c-424e-b9b1-ac98582adbd8.gif)

**Reply by lava83** [reply]
Good Morning, 

we use Livewire 2.7.2. 

![Bildschirmfoto 2021-11-18 um 07 59 31](https://user-images.githubusercontent.com/16590655/142367687-a8cb5c26-1257-4d7f-a031-84bdf075a405.png)

I've also added the `$errors` variables and `@error` - directive inside a separate div element but cant see any errors.

**Reply by lava83** [reply]
I have now put the component into the default namespace: `App\Http\Livewire` - and the view inside: `resources/views/livewire` then the error messages work for me. If I use the component within another namespace like: `App\FooBar\Livewire\Components`, it does not work anymore. 

The `serverMemo` object has an `errors` object, but the errors are not displayed in the component, under its own namespace. 

### OwnNameSpace
![livewire-playground-own-namesapce](https://user-images.githubusercontent.com/16590655/142374243-0f332c3c-456a-4d66-846c-0346d3f7b4c8.gif)

### DefaultNameSpace
![livewire-playground-default-namesapce](https://user-images.githubusercontent.com/16590655/142374279-ad385f1b-91ca-4859-a0e2-04de9191e2f3.gif)

**Reply by zepfietje** [reply]
Interesting. Not sure why it happens, but you might want to look here:
https://github.com/livewire/livewire/blob/6ee7e219f2d568b57a657398f59a19ff17f91d29/src/Component.php#L198

**Reply by lava83** [reply]
I've added a `dump($view)` - in the method and see, if I use the default namespace, the view object looks like:

```php
Illuminate\View\View {#393 ▼
  #factory: Illuminate\View\Factory {#369 ▶}
  #engine: Livewire\CompilerEngineForIgnition {#365 ▶}
  #view: "livewire.import-stock-items-component"
  #data: array:3 [▶]
  #path: "/Users/stefan/Code/livewireplayground/resources/views/livewire/import-stock-items-component.blade.php"
}
```

here you can see the `$view->view` property is: `livewire.import-stock-items-component` 

But when I use the other namespace it looks like:

```php
Illuminate\View\View {#406 ▼
  #factory: Illuminate\View\Factory {#369 ▶}
  #engine: Livewire\CompilerEngineForIgnition {#365 ▶}
  #view: "__components::79bebd8d59d7ad1495b159b0509af0f1e72252df"
  #data: array:4 [▶]
  #path: "/Users/stefan/Code/livewireplayground/storage/framework/views/79bebd8d59d7ad1495b159b0509af0f1e72252df.blade.php"
}
```

now you can see, Laravel will use a cached Blade Template. The content of: `/Users/stefan/Code/livewireplayground/storage/framework/views/79bebd8d59d7ad1495b159b0509af0f1e72252df.blade.php` is: 

```blade
<div>
    Errors:
    <div>[]</div>
    <input type="file" wire:model="importFile">
    <div></div>
    <button wire:click="import">Import</button>
</div>
```
![Bildschirmfoto 2021-11-18 um 09 59 04](https://user-images.githubusercontent.com/16590655/142383955-20842e17-222a-4ae0-ac4d-1e4429754f3a.png)

And if I will go deeper to: `Illuminate\View\View@render` and add a `dump($this)` I can see the following by use the default namespace:

![Bildschirmfoto 2021-11-18 um 10 07 11](https://user-images.githubusercontent.com/16590655/142384984-1c23702f-0876-4b7c-9b9e-1f15de891564.png)

the render method will call two times and every time with view: `"livewire.import-stock-items-component"`. But if I use my namespace you will see this:

![Bildschirmfoto 2021-11-18 um 10 09 07](https://user-images.githubusercontent.com/16590655/142385297-688a691a-6873-4794-9659-93cd0a695a4d.png)

Here the render method will call three times(?) - and the first call is with the correct view but the second last are called with the cached(?) blade template.

Im very confused..

**Reply by zepfietje** [reply]
Hmm really interesting.

Is the Livewire component a full-page component?

**Reply by lava83** [reply]
> Hmm really interesting.
> 
> Is the Livewire component a full-page component?

No stand alone. And will fires with:

```blade
@livewire('srit83::import-foobar')
```

**Reply by joshhanley** [reply]
When you are moving your component to a custom directory structure, are you registering the component in a Service Provider? If so, can you share that code?

Also when you move the class, are you also moving the view? If so what is the full path of the view in the custom location?

**Reply by lava83** [reply]
> When you are moving your component to a custom directory structure, are you registering the component in a Service Provider? If so, can you share that code?

Yes, here:

```php
use App\Srit83\Livewire\Components\ImportStockItemsComponent;
use Illuminate\Support\ServiceProvider;
use Livewire\Livewire;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        Livewire::component('srit83::import-foobar', ImportStockItemsComponent::class);
    }
...
```

> Also when you move the class, are you also moving the view? If so what is the full path of the view in the custom location?

The component class path is: `app/Srit83/Livewire/Components/ImportStockItemsComponent.php` and the view path is: `app/Srit83/Livewire/Views/import.blade.php` - and the view will returned in render like this: 

```php
public function render(): string
{
    return view('Srit83.Livewire.Views.import');
}
```

**Reply by lava83** [reply]
And in: `config/view.php` - I will set the view path like this:

```php
'paths' => [
    resource_path('views'),
    app_path(),
],
```

**Reply by joshhanley** [reply]
@lava83 hmm maybe it is still picking up an old compiled version of it from when it was in the resources directory. Have you tried running `php artisan view:clear` to clear all cached blade views?

**Reply by lava83** [reply]
> @lava83 hmm maybe it is still picking up an old compiled version of it from when it was in the resources directory. Have you tried running `php artisan view:clear` to clear all cached blade views?

Yeah, I happened to do a `view:clear` yesterday. And after that, the mission was complete. ;-) Thanks anyway.
